### PR TITLE
[TU-35] The toggle component

### PR DESCRIPTION
**Pull request by lowiebenoot**
screen shot 2017-10-30 at 16 00 46

screen shot 2017-10-30 at 16 00 37

screen shot 2017-10-30 at 16 01 37

### CODE REVIEW

File: components/identifiers.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export const BUTTON = 'TLButton';
+export const DIALOG = 'TLDialog';
+export const LOADING_MOLECULE = 'TLLoadingMolecule';
+export const MENU = 'TLMenu';
+export const OVERLAY = 'TLOverlay';
+export const POPOVER_VERTICAL = 'TLPopoverVertical';
+export const POPOVER_HORIZONTAL = 'TLPopoverHorizontal';
+export const RADIO = 'TLRadio';
+export const TOAST = 'TLToast';
+export const TOGGLE = 'TLToggle';
```
Comment (nickwaelkens) [inline components/identifiers.js]: Legacy from `react-themr`? 🔥

File: components/toggle/Toggle.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    size: 'medium',
+    color: 'neutral',
+  };
+
+  constructor() {
+    super(...arguments);
+    this.handleToggle = ::this.handleToggle;
+  }
+
+  handleToggle() {
```
Comment (nickwaelkens) [inline components/toggle/Toggle.js]: Support for `redux-form` 👏

File: components/toggle/theme.css [hunk 3]
```diff
@@ -0,0 +1,363 @@
+@import 'variables.css';
+@import '@teamleader/ui-colors/index.css';
```
Comment (nickwaelkens) [inline components/toggle/theme.css]: `index.css` is implied in the `ui-colors` `package.json` file, can be removed 👍

File: components/toggle/theme.css [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  --toggle-thumb-width-medium: 18px;
+
+  --toggle-track-height-large: 30px;
+  --toggle-track-width-large: 54px;
+  --toggle-thumb-width-large: 24px;
+
+  --toggle-track-padding: 3px;
+}
+
+.toggle {
```
Comment (nickwaelkens) [inline components/toggle/theme.css]: Can we not rewrite this using parent selectors? Or do we need another PostCSS plugin for that 🤔 

```css
.thumb {
  .medium & {
    /* whatever */
}
```

File: components/identifiers.js (hunk 1 above)
Comment (lowiebenoot) [reply components/identifiers.js]: Damn yes, should be gone now

File: components/toggle/theme.css (hunk 4 above)
Comment (lowiebenoot) [reply components/toggle/theme.css]: Yes, it works with `postcss-nested`. Going to refactor my css tomorrow.

File: components/toggle/Toggle.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    const classNames = cx(theme['toggle'], className, theme[size], theme[color], {
+      [theme['checked']]: checked,
+      [theme['disabled']]: disabled,
+    });
+
+    return (
+      <label data-teamleader-ui="toggle" className={classNames}>
+        <input
+          {...rest}
+          className={theme.input}
```
Comment (driesd) [inline components/toggle/Toggle.js]: This should be converted to the array notation: theme['input']

File: components/toggle/Toggle.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+          type="checkbox"
+          checked={checked}
+          onClick={this.handleToggle}
+          readOnly
+        />
+        <span className={theme.track}>
```
Comment (driesd) [inline components/toggle/Toggle.js]: This should be converted to the array notation: theme['track']

File: components/toggle/Toggle.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+          <span className={theme.thumb} />
```
Comment (driesd) [inline components/toggle/Toggle.js]: This should be converted to the array notation: theme['thumb']

File: components/toggle/Toggle.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    if (!this.props.disabled && this.props.onChange) {
+      this.props.onChange(!this.props.checked, event);
+    }
+  }
+
+  render() {
+    const { checked, disabled, className, size, color, ...others } = this.props;
+    const rest = omit(others, ['onChange']);
+
+    const classNames = cx(theme['toggle'], className, theme[size], theme[color], {
```
Comment (driesd) [inline components/toggle/Toggle.js]: I think that **className** should be the last one in row to allow overriding

File: components/toggle/Toggle.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  };
+
+  constructor() {
+    super(...arguments);
+    this.handleToggle = ::this.handleToggle;
+  }
+
+  handleToggle() {
+    if (!this.props.disabled && this.props.onChange) {
+      this.props.onChange(!this.props.checked, event);
```
Comment (driesd) [inline components/toggle/Toggle.js]: PP: Maybe you can extract **checked, disabled, onChange** from this.props?

File: components/toggle/Toggle.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    const { checked, disabled, className, size, color, ...others } = this.props;
+    const rest = omit(others, ['onChange']);
+
+    const classNames = cx(theme['toggle'], className, theme[size], theme[color], {
+      [theme['checked']]: checked,
+      [theme['disabled']]: disabled,
+    });
+
+    return (
+      <label data-teamleader-ui="toggle" className={classNames}>
```
Comment (driesd) [inline components/toggle/Toggle.js]: The component is wrapped in a **label** tag, but I don't see any label prop being rendered?

File: components/toggle/theme.css [hunk 11]
```diff
@@ -0,0 +1,363 @@
+@import 'variables.css';
```
Comment (driesd) [inline components/toggle/theme.css]: You need to import **variables** here, since you're not using any from this file.

**Comment by lowiebenoot** [comment]
fyi: it's still WIP

File: components/toggle/Toggle.js (hunk 8 above)
Comment (lowiebenoot) [reply components/toggle/Toggle.js]: Good catch 👍

File: components/toggle/Toggle.js (hunk 10 above)
Comment (lowiebenoot) [reply components/toggle/Toggle.js]: Well, it's a label to make "invisible checkbox" work. It's just a small trick.

Comment (driesd) [reply components/toggle/Toggle.js]: Ok, good stuff! But I think we'll need a label (rendered next to the toggle) as well?

schermafdruk 2017-10-19 09 35 28

Comment (lowiebenoot) [reply components/toggle/Toggle.js]: yes, I just still need to add it

File: components/toggle/theme.css (hunk 4 above)
Comment (lowiebenoot) [reply components/toggle/theme.css]: Tried to refactor with using parent selectors but it didn't really work out. A lot of things weren't possible or were giving me css specificity issues.

 :ok: by @lowiebenoot on 19/10/2017 11:51:53

File: components/toggle/Toggle.js (hunk 10 above)
Comment (lowiebenoot) [reply components/toggle/Toggle.js]: :ok: by @lowiebenoot on 26/10/2017 10:22:53

**Comment by lowiebenoot** [comment]
Toggle has been finalised, please review!

**Comment by nickwaelkens** [comment]
Is this ready to be merged or are there any styleguide changes to be implemented here too? (cfr. stand up)

**Comment by lowiebenoot** [comment]
I don't think it needs changes, I asked Stephen to confirm. But in the meanwhile, you can still review 👍

**Comment by lowiebenoot** [comment]
Stephen confirmed, no changed needed.

**Pull request merged**